docs/Learning-Environment-Design-Agents.md

```diff
@@ -927,6 +929,91 @@ provide examples of symmetric games. To train an asymmetric game, specify
+
+Cooperative behavior in ML-Agents can be enabled by instantiating a `SimpleMultiAgentGroup`,
+typically in an environment controller or similar script, and adding agents to it
+using the `RegisterAgent(Agent agent)` method. Using `MultiAgentGroup` enables the
+agents within a group to learn how to work together to achieve a common goal (i.e.,
+maximize a group-given reward), even if one or more of the group members are removed
+before the episode ends. You can then use this group to add/set rewards, end or interrupt episodes
+at a group level using the `AddGroupReward()`, `SetGroupReward()`, `EndGroupEpisode()`, and
+`GroupEpisodeInterrupted()` methods. For example:
+
+```csharp
+// Create a Multi Agent Group in Start() or Initialize()
+m_AgentGroup = new SimpleMultiAgentGroup();
+
+// Register agents in group at the beginning of an episode
+for (var agent in AgentList)
+{
+  m_AgentGroup.RegisterAgent(agent);
+}
+
+// if the team scores a goal
+m_AgentGroup.AddGroupReward(score);
+
+// If the goal is reached and the episode is over
+m_AgentGroup.EndGroupEpisode();
+ResetScene();
+
+// If time ran out and we need to interrupt the episode
+m_AgentGroup.GroupEpisodeInterrupted();
+ResetScene();
+```
+
+Multi Agent Groups are best used with the COMA2 trainer, which will train a "coach" for each
+Group. This can be enabled by using the `coma` trainer - see the
+[training configurations](Training-Configuration-File.md) doc for more information on
+configuring COMA2. When using COMA2, agents which are deactivated or removed from the Scene
+during the episode will still "receive" group rewards that are seen later.
+
+**NOTE**: Groups differ from Teams (for competitive settings) in the following way - agents
+working together should be added to the same Group, while Agents playing against each other
+should be given different Team Ids. If in the Scene there is one playing field and two teams,
+there should be two Groups, one for each team, and each team should be assigned a different
+Team Id. If this playing field is duplicated many times in the Scene (e.g. for training
+speedup), there should be two Groups _per playing field_, and two unique Team Ids
+_for the entire Scene_.
+
+For an example of how to set up cooperative environments, see the
+[Cooperative PushBlock](Learning-Environment-Examples.md#cooperative-push-block) and
+Escape Room example environments.
+
+#### Cooperative Behaviors Notes and Best Practices
+* An Agent can only be registered to one MultiAgentGroup at a time. If you want to re-assign an
+Agent from one group to another, you have to unregister it from the current group first.
+
+* Agents with different behavior names in the same group are not supported.
+
+* Agents within groups should always set the `Max Steps` parameter the Agent script to 0, meaning
+they will never reach a max step. Instead, handle Max Steps with MultiAgentGroup by ending the episode for the entire
+Group using `GroupEpisodeInterrupted()`.
```

Comment:
Or EndGroupEpisode?

Reply:
here I want to emphasize how to handle max step reached and for max step we should use GroupEpisodeInterrupted().